---
params:
    root_directory: NULL
    input_seurat_object: NULL
    output_seurat_object: NULL
title: "NBIS Support #5568"
subtitle: "Quality controls and filtering"
author: "Erik Fasterius"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float:
            collapsed: true
---

```{r Setup}
# Knit options
knitr::opts_knit$set(root.dir = params$root_directory)

# Chunk options
knitr::opts_chunk$set(fig.width  = 12,
                      fig.height = 8,
                      fig.align  = "center",
                      message    = FALSE,
                      warning    = FALSE,
                      results    = "asis")
```

```{r Load packages}
suppressPackageStartupMessages({
    library("cowplot")
    library("tximport")
    library("Seurat")
    library("tidyverse")
})
```

```{r Read data}
s_obj <- readRDS(params$input_seurat_object)
```

```{r Plotting functions}
# Function for plotting basic QC metrics
plot_qc_metrics <- function(s_obj) {
    features <- c("nFeature_RNA", "nCount_RNA", "percent_mito", "percent_ribo",
                  "S.Score", "G2M.Score")
    gg <- VlnPlot(s_obj,
                  group.by = "orig.ident",
                  features = features,
                  pt.size  = 0.1,
                  ncol     = 2) +
        NoLegend()
    show(gg)
}

# Function for plotting features versus counts
plot_features_vs_counts <- function(s_obj) {
    gg <- FeatureScatter(s_obj,
                         "nCount_RNA",
                         "nFeature_RNA",
                         group.by = "orig.ident",
                         pt.size  = 0.5) +
        theme(plot.title = element_text(hjust = 0.5)) +
        ggtitle("Counts vs. features")
    show(gg)
}

# Function for plotting the top 20 highest variable features
plot_hvg <- function(s_obj) {
    s_obj_tmp <- FindVariableFeatures(s_obj,
                                      selection.method = "vst",
                                      nfeatures        = 20)
    top_20 <- head(VariableFeatures(s_obj_tmp), 20)
    gg <- LabelPoints(plot   = VariableFeaturePlot(s_obj_tmp),
                      points = top_20,
                      repel  = TRUE) +
        theme(plot.title = element_text(hjust = 0.5)) +
        ggtitle("Top 20 variable features")
    show(gg)
}

# Function for plotting the top 20 percentage counts per gene
plot_highest_percentages <- function(s_obj) {
    counts <- s_obj@assays$RNA@counts
    counts <- Matrix::t(Matrix::t(counts) / Matrix::colSums(counts)) * 100
    most_expressed <- order(apply(counts, 1, median), decreasing = TRUE)[20:1]
    most_expressed <- as.matrix(Matrix::t(counts[most_expressed, ]))
    par(mar = c(4, 10, 2, 1))
    boxplot(most_expressed,
            cex  = 0.1,
            las  = 1,
            xlab = "% total count per cell",
            col  = scales::hue_pal()(20)[20:1], horizontal = TRUE)
}
```

# Unfiltered data

The following plots are performed on the raw, unfiltered expression data, with
the sole exception of reading in the data itself, which has been filtered to
exclude features (*i.e.* genes or transcripts, depending on the available
annotation) that are expressed in fewer than 3 cells, as well as excluding
cells which have fewer than 10 features.

## Quality controls

```{r Normalisation and basic quality controls}
# Normalise data
s_obj <- NormalizeData(s_obj,
                       normalization.method = "LogNormalize",
                       scale.factor         = 10000)

# Add mitochondrial / ribosomal gene percentages
s_obj[["percent_mito"]] <- PercentageFeatureSet(s_obj, pattern = "^MT-")
s_obj[["percent_ribo"]] <- PercentageFeatureSet(s_obj, pattern = "^RP[SL]")

# Cell cycle scoring
s_obj <- CellCycleScoring(s_obj,
                          g2m.features = cc.genes$g2m.genes,
                          s.features   = cc.genes$s.genes)
```

The first plot shows the distribution of the basic QC metrics across the two
samples, here names `S1` and `S2`. There are fewer features expressed per cell
in the first sample compared to the second (`nFeature_RNA`), while the first
sample has a higher number of reads per feature (`nCount_RNA`). The proportion
of mitochondrial reads (`percent_mito`) are relatively large, especially for
the second sample, while the proportion of ribosomal reads (`percent_ribo`) are
more reasonable (though sample two is still higher). Cell cycle scores (based
on the standard human gene names; `S.Score` and `G2M.Score`) are similar for
the two samples.

```{r Plot basic QC metrics}
plot_qc_metrics(s_obj)
```

The second plot shows a similar thing as the above regarding counts and
features: the first sample seems to be of generally lower quality with a
potentially less complex library (fewer features with more counts per feature).

```{r Features vs. counts}
plot_features_vs_counts(s_obj)
```

The third plot just highlights the top 20 most highly variable genes:

```{r Plot highest variable genes}
plot_hvg(s_obj)
```

The fourth plot shows the genes with the highest proportion of counts per cell.
Some of these are mitochondrial and ribosomal genes, indicating possible low
quality cells. Other genes of interest are `pol`, `NIrc3`, `His3` and the
un-annotated `comp36811-c0-seq6` transcript. If these features are not of
interest we could consider removing them from the subsequent analyses.

```{r Plot highest counts per cell}
plot_highest_percentages(s_obj)
```

## Clustering

We now cluster the unfiltered cells, to give us a baseline to work with. The
first figure shows the clustering with the cells coloured by their sample,
showing that the largest driving factor of the clustering here seems to be the
sample of origin. There are some cells from both samples that cluster with the
other, however. The second figure shows the clustering coloured by the clusters
themselves, all 17 of them.

We then want to see if any of the clusters are driven by any of the QC metrics
we already investigated above - this is what the last plot (with six panels)
show. While there is some variation in how the various QC metrics are
distributed across the cells and clusters, none of them stand out as a
particular driver of the clustering. This may be because that the effect of the
samples overcome any effect the QC metrics (*i.e.* cell quality) may have on
the clustering in this particular instance.

```{r Variable genes, scaling and clustering, fig.height = 16}
# Function for trying out different number of variable genes
cluster_cells <- function(s_obj,
                          nfeatures  = 3000,
                          npcs       = 100,
                          dims       = 1:50,
                          resolution = 0.8) {

    # Find the highest variable features
    s_obj <- FindVariableFeatures(s_obj,
                                  selection.method = "vst",
                                  nfeatures        = nfeatures)

    # Scale data
    to_regress <- c("percent_mito", "nFeature_RNA", "S.Score", "G2M.Score")
    s_obj <- ScaleData(s_obj,
                       assay           = "RNA",
                       vars.to.regress = to_regress)

    # Dimensionality reduction and clustering
    s_obj <- RunPCA(s_obj,
                    npcs     = npcs,
                    features = VariableFeatures(object = s_obj))
    s_obj <- FindNeighbors(s_obj,
                           reduction = "pca",
                           dims      = dims,
                           verbose   = FALSE)
    s_obj <- FindClusters(s_obj,
                          reduction  = "pca",
                          resolution = resolution,
                          verbose    = FALSE)

    # Run and plot UMAP
    s_obj <- RunUMAP(s_obj,
                     dims                 = dims,
                     min.dist             = 0.01,
                     spread               = 0.3,
                     metric               = "cosine",
                     n.neighbors          = 15,
                     repulsion.strength   = 0.5,
                     negative.sample.rate = 20,
                     n.epochs             = 100,
                     verbose              = FALSE)
    gg1 <- DimPlot(s_obj, reduction = "umap", group.by = "orig.ident") +
        ggtitle("Samples")
    gg2 <- DimPlot(s_obj, reduction = "umap", group.by = "seurat_clusters") +
        ggtitle("Clusters")
    gg <- plot_grid(gg1, gg2, ncol = 1)
    show(gg)

    # Plot selected features on UMAP
    features <- c("nCount_RNA", "nFeature_RNA", "percent_mito",
                  "percent_ribo", "S.Score", "G2M.Score")
    gg <- FeaturePlot(s_obj, reduction = "umap", features = features)
    show(gg)

    # Return Seurat object
    return(s_obj)
}

# Perform clustering
s_obj <- cluster_cells(s_obj, nfeatures = 2000)
```

```{r Save to RDS}
saveRDS(s_obj, file = params$output_seurat_object)
```
