---
params:
    root_directory: NULL
    input_seurat_object: NULL
title: "NBIS Support #5568"
subtitle: "Differential expression analyses"
author: "Erik Fasterius"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float:
            collapsed: true
---

```{r Setup}
# Knit options
knitr::opts_knit$set(root.dir = params$root_directory)

# Chunk options
knitr::opts_chunk$set(fig.width  = 12,
                      fig.height = 8,
                      fig.align  = "center",
                      message    = FALSE,
                      warning    = FALSE,
                      results    = "asis")
```

```{r Load packages}
suppressPackageStartupMessages({
    library("cowplot")
    library("ggplot2")
    library("MAST")
    library("Seurat")
    library("tidyverse")
})
```

```{r Read data}
s_obj <- readRDS(params$input_seurat_object)
```

# Differential expression analyses

Perform differential expression comparing each individual cluster to all other
clusters, yielding features (genes or transcripts, in this case) that are
differentially expressed for each particular cluster compared to the rest. Plot
the top nine differentially expressed features for each comparison; the full
list of all features is included in the output files on the format
`degs.<cluster>-vs-all.tsv`.

```{r Differential expression in all clusters}
markers <- FindAllMarkers(object              = s_obj,
                          assay               = "RNA",
                          test.use            = "MAST",
                          slot                = "data",
                          latent.vars         = NULL,
                          min.pct             = 0.1,
                          logfc.threshold     = 0.25,
                          return.thresh       = 0.01)

# Re-order columns
markers <- markers %>%
    select(cluster, gene, everything())

# Find top 10 DEGs per cluster
top_9 <- markers %>%
    group_by(cluster) %>%
    top_n(-9, p_val_adj) %>%
    pull(gene) %>%
    unique() %>%
    gsub("\\\\","-", .)
```

## Heatmap

```{r Heatmap}
DoHeatmap(object   = s_obj,
          features = top_9,
          slot     = "data",
          group.by = "seurat_clusters",
          assay    = "RNA")
```

## Dotplot

```{r Dotplot, fig.height = 24}
DotPlot(object   = s_obj,
        features = top_9,
        group.by = "seurat_clusters",
        assay    = "RNA") +
    coord_flip()
```

## Violinplot

```{r Violinplot, fig.height = 24}
VlnPlot(object   = s_obj,
        features = top_9,
        ncol     = 9,
        group.by = "seurat_clusters",
        assay    = "RNA")
```

## Expression per cluster

```{r Plot clusters}
DimPlot(s_obj,
        reduction = "umap",
        group.by  = "seurat_clusters",
        label     = TRUE) +
    NoAxes() +
    NoLegend() +
    ggtitle("Clusters")
```

```{r Plot top 9 features per cluster}
# Loop over all individual clusters
for (cluster in sort(unique(markers$cluster))) {

    # Print cluster header
    cat(paste0("\n\n### Cluster ", cluster, "\n\n"))

    # Get DEGs for current cluster
    current_markers <- markers[markers$cluster == cluster, ]

    # Write current DEGs to file
    file <- paste0("degs.cluster-", cluster, "-vs-all.tsv")
    write.table(current_markers,
                file      = file,
                sep       = "\t",
                row.names = FALSE,
                quote     = FALSE)

    # Plot top 9 features
    top_9 <- as.character(current_markers$gene[1:9])
    plot_list <- FeaturePlot(s_obj,
                             features = top_9,
                             combine  = FALSE)

    # Remove all axes and legends
    for(ii in 1:length(plot_list)) {
        plot_list[[ii]] <- plot_list[[ii]] +
            NoLegend() +
            NoAxes()
    }

    # Plot in a grid
    show(cowplot::plot_grid(plotlist = plot_list))
}
```
