---
params:
    root_directory: NULL
    input_seurat_object: NULL
title: "NBIS Support #5568"
subtitle: "Differential expression analyses"
author: "Erik Fasterius"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float:
            collapsed: true
---

```{r Setup}
# Knit options
knitr::opts_knit$set(root.dir = params$root_directory)

# Chunk options
knitr::opts_chunk$set(fig.width  = 12,
                      fig.height = 8,
                      fig.align  = "center",
                      message    = FALSE,
                      warning    = FALSE,
                      results    = "asis")

# Set seed
set.seed(42)
```

```{r Load packages}
suppressPackageStartupMessages({
    library("cowplot")
    library("ggplot2")
    library("MAST")
    library("Seurat")
    library("tidyverse")
})
```

# Differential expression analyses

Perform differential expression comparing each individual cluster to all other
clusters, yielding features (genes or transcripts, in this case) that are
differentially expressed for each particular cluster compared to the rest. Plot
the top nine differentially expressed features for each comparison; the full
list of all features is included in the output files on the format
`degs.<cluster>-vs-all.tsv`.

```{r Read data}
s_obj <- readRDS(params$input_seurat_object)
```

```{r Differential expression in all clusters}
# Perform DEA
markers <- FindAllMarkers(object              = s_obj,
                          assay               = "RNA",
                          test.use            = "MAST",
                          slot                = "data",
                          latent.vars         = NULL,
                          min.pct             = 0.1,
                          logfc.threshold     = 0.25,
                          return.thresh       = 0.01)

# Re-order columns
markers <- markers %>%
    select(cluster, gene, everything())

# Find top 10 DEGs per cluster
top_9 <- markers %>%
    group_by(cluster) %>%
    top_n(-9, p_val_adj) %>%
    pull(gene) %>%
    unique() %>%
    gsub("\\\\","-", .)

# Plot cluster IDs
DimPlot(s_obj,
        reduction = "umap",
        group.by  = "seurat_clusters",
        label     = TRUE) +
    NoAxes() +
    NoLegend() +
    ggtitle("Clusters")
```

## Heatmap

The following is a heatmap for the top nine differentially expressed features in
each of the clusters. The top features for each cluster is included, which is
why many more than nine are shown in total. We can see that the genes that are
included generally have a higher expression than individual transcripts, thus
contributing more to the clustering of the heatmap.

```{r Heatmap, fig.height = 32}
DoHeatmap(object   = s_obj,
          features = top_9,
          slot     = "data",
          group.by = "seurat_clusters",
          assay    = "RNA")
```

## Dotplot

The following is a dotplot for the top nine differentially expressed features in
each of the clusters, visualising the expression difference for each feature
between the clusters of the data.

```{r Dotplot, fig.height = 32}
DotPlot(object   = s_obj,
        features = top_9,
        group.by = "seurat_clusters",
        assay    = "RNA") +
    coord_flip()
```

## Violinplots

The following are violinplots of the top nine differentially expressed features
of the data, which is another way to highlight differences of feature expression
between the clusters of the data. Expression is low overall, but a closer view
of any particular feature shows in which clusters that particular feature is
expressed. There are not many features that are uniquely expressed in a single
cluster, but some exist (in particular for cluster 15).

```{r Violinplot, fig.height = 60, fig.width = 20}
VlnPlot(object   = s_obj,
        features = top_9,
        ncol     = 5,
        group.by = "seurat_clusters",
        assay    = "RNA")
```

## Expression per cluster

The following figures show the expression of the top 9 differentially expressed
genes for each cluster in the UMAP of the data, highlighting the location of
feature expression in UMAP-space. This is another way to highlight feature
expression in different clusters and what clusters are similar to each other in
regards to expression of each cluster's most unique features.

```{r Plot top 9 features per cluster}
# Plot cluster IDs
DimPlot(s_obj,
        reduction = "umap",
        group.by  = "seurat_clusters",
        label     = TRUE) +
    NoAxes() +
    NoLegend() +
    ggtitle("Clusters")

# Loop over all individual clusters
for (cluster in sort(unique(markers$cluster))) {

    # Print cluster header
    cat(paste0("\n\n### Cluster ", cluster, "\n\n"))

    # Get DEGs for current cluster
    current_markers <- markers[markers$cluster == cluster, ]

    # Write current DEGs to file
    file <- paste0("degs.cluster-", cluster, "-vs-all.tsv")
    write.table(current_markers,
                file      = file,
                sep       = "\t",
                row.names = FALSE,
                quote     = FALSE)

    # Plot top 9 features
    top_9 <- as.character(current_markers$gene[1:9])
    plot_list <- FeaturePlot(s_obj,
                             features = top_9,
                             order    = TRUE,
                             combine  = FALSE)

    # Remove all axes and legends
    for(ii in 1:length(plot_list)) {
        plot_list[[ii]] <- plot_list[[ii]] +
            NoLegend() +
            NoAxes()
    }

    # Plot in a grid
    show(cowplot::plot_grid(plotlist = plot_list))
}
```
